kclvm/tools/src/LSP/src/completion.rs

```diff
@@ -142,34 +107,119 @@ pub(crate) fn completion(
-            Stmt::Import(stmt) => completion_for_import(&stmt, pos, prog_scope, program),
-            _ => Some(into_completion_items(&get_dot_completion(node, pos, prog_scope)).into()),
-        },
-        None => None,
+    if let Some(stmt) = program.pos_to_stmt(&pos) {
+        match stmt.node {
+            Stmt::Import(stmt) => return completion_import(&stmt, &pos, prog_scope, program),
+            _ => {
+                let (expr, parent) = inner_most_expr_in_stmt(&stmt.node, &pos, None);
+                if let Some(node) = expr {
+                    if let Expr::StringLit(s) = node.node {
+                        return Some(
+                            into_completion_items(
+                                &STRING_MEMBER_FUNCTIONS
+                                    .iter()
+                                    .map(|(name, ty)| KCLCompletionItem {
+                                        label: func_ty_complete_label(name, &ty.into_function_ty()),
+                                        detail: Some(ty.ty_str()),
+                                        documentation: ty.ty_doc(),
+                                        kind: Some(KCLCompletionItemKind::Function),
+                                    })
+                                    .collect(),
+                            )
+                            .into(),
+                        );
+                    }
+                }
+            }
+        }
     }
+
+    let mut def = find_def_with_gs(&pos, &gs, true);
+    if def.is_none() {
+        let pos = KCLPos {
+            filename: pos.filename.clone(),
+            line: pos.line,
+            column: pos.column.map(|c| c + 1),
+        };
+        def = find_def_with_gs(&pos, &gs, false);
+    }
+    match def {
+        Some(def_ref) => {
+            if let Some(def) = gs.get_symbols().get_symbol(def_ref) {
+                let module_info = gs.get_packages().get_module_info(&pos.filename);
+                let attrs = def.get_all_attributes(gs.get_symbols(), module_info);
+                for attr in attrs {
+                    let attr_def = gs.get_symbols().get_symbol(attr);
+                    if let Some(attr_def) = attr_def {
+                        let sema_info = attr_def.get_sema_info();
+                        let name = attr_def.get_name();
+                        match &sema_info.ty {
+                            Some(ty) => {
+                                let label: String = match &ty.kind {
+                                    kclvm_sema::ty::TypeKind::Schema(schema) => {
+                                        schema_ty_completion_item(schema).label
+                                    }
+                                    kclvm_sema::ty::TypeKind::Function(func_ty) => {
+                                        func_ty_complete_label(&name, &func_ty)
+                                    }
+                                    _ => name.clone(),
```

Comment:
Do we need to handle string types here? And right How is the handling of `?.` operators reflected in the code?

Reply:
The type here is the type that needs to be completed (that is, the type of the value on the right side of `.`)
I guess what you want to ask is about completion of `"s".(lower())`, where the builtin function of string type will be added to `attrs`, and here it will be treated as a function type
The appearance of the string type here indicates that a string type attr needs to be completed, such as `Person{}.(name)`